### DISCUSSION TITLE
Getting 413 response from getStaticProps Vercel serverless function with 750kb response payload

### DISCUSSION THREAD

**Discussion by MichaelBurgess** [vercel/next.js#37466] [Help]
Hi - I'm pretty stuck on this one...I've got a number of SSG pages (that use ISR to regenerate) that are returning a 413 error in Vercel from the serverless function that backs `getStaticProps`:

`LAMBDA_RUNTIME Failed to post handler success response. Http response code: 413.`

I've got a log inside that function with the size of the payload that's called right before I return it and I know that the pages in question are between 750-780kb, but I'm still getting this 413 error.

This only happens when deployed to production in Vercel and only when you navigate directly to the page in the browser, not when you do client-side routing from other pages to an affected page.

I'm pretty baffled as I'm not even anywhere near the 5MB limit (not even 20% of it), but I'm getting this? The [docs](https://vercel.com/docs/concepts/limits/overview#serverless-function-payload-size-limit) say this is for request body or response body - I've validated that it's not the response body, how can I check the request body? I don't control this function - it's generated automatically for me and all my `getStaticPaths` returns for each path are the 3 strings needed to build the dynamic path, so we're talking nothing at all on the way in either.

The closest I've found to this issue on here is someone who had 4MB payload and the marked answer referred to the docs saying the limit is 5MB, which makes no sense as the OP was at or under 4MB. In my case, I'm way under. I can't find the link to that item now unfortunately...

If I reduce the payload down to something small like 10kb by removing some things (that I actually need still - just for debugging purposes), the error goes away - so it feels like the limit isn't actually 5MB in reality if 750kb triggers this error?

Any ideas how I can get to the bottom of what's going on? Is there some other payload added in that I can’t control?

Thanks!

Next.js 12.1.6.

**Comment by icyJoseph** [comment]
Hi,

Some places to keep on searching:

It is not only space, but also time that's limited in these lambda runtimes, I think the timeout happens at 10 seconds, after which their task runner kills the lambda. 

Going about space again, perhaps the lambda is using too much RAM when processing the data?

> This only happens when deployed to production in Vercel and only when you navigate directly to the page in the browser, not when you do client-side routing from other pages to an affected page.

This is very strange to me. Pages with `getStaticProps` are made at build time, so that they are served instantly at runtime, why would a lambda have to run then, unless you are using preview mode?

It could also be that you have tons of cookies? JWT's grow large and reach easily 8MB.

**Reply by MichaelBurgess** [reply]
Thanks for the reply @icyJoseph.

We're on the team/pro package, so I believe the timeout is 1 minute? Not cookies also in this instance - good suggestion though!

Does a serverless function run to serve up the initial page to you? Could that be a factor here? There's a lot of data needed to initially populate the page, so perhaps there's some request handler running that goes over 5mb?

There must be some function running as I can watch the serverless function logs for my deployment and see the failure happen in realtime as I try to load the initial page.

`getStaticProps` is also called after a period of time at runtime when using ISR right? (we're using ISR). But I know that payload is max 780kb so I can't see how it's that.

I'm a little stumped TBH - I know that I need to generally try to reduce the size of data I load, or perhaps when I load it (will open a diff issue as I have a qu on that), but I'm a little baffled by this one...

**Reply by icyJoseph** [reply]
> We're on the team/pro package, so I believe the timeout is 1 minute? Not cookies also in this instance - good suggestion though!

Perhaps you need to configure that on the .vercel file? I am not sure if the max is activated by default. 
 
> Does a serverless function run to serve up the initial page to you? Could that be a factor here? There's a lot of data needed to initially populate the page, so perhaps there's some request handler running that goes over 5mb?

Technically, I can't answer for sure 100%, but pages made with `getStaticProps` exist as `html` so the server just sends that to the browser on first request. To make it `ultra fast` as they say.

> There must be some function running as I can watch the serverless function logs for my deployment and see the failure happen in realtime as I try to load the initial page.

Right, the server does most likely have a function that serves the files needed for that page. 

Could you try and see the total size of the page? Maybe with all of the HTML and JSON payloads, it goes over 5MB? That'd be a strange limitation though.

Build it locally, and use any tool to see the size of the pages in `.next/pages/server`

> `getStaticProps` is also called after a period of time at runtime when using ISR right? (we're using ISR). But I know that payload is max 780kb so I can't see how it's that.

Yeah, but ISR runs some time after the visit, not as you visit the page, and the user wouldn't even notice. Btw, this ends up as a blank screen with the error, right? or do you still serve the page generated at build time?

> I'm a little stumped TBH - I know that I need to generally try to reduce the size of data I load, or perhaps when I load it (will open a diff issue as I have a qu on that), but I'm a little baffled by this one...

Yeah, well some data could be fetched on the client, though it might hurt SEO if its super important.

**Reply by MichaelBurgess** [reply]
I think you're onto something with doing the local build - I just generated the offending pages locally with a build and they are around 7.5 meg. I'll dig into what on earth is in them to make them so large....thanks!

**Reply by pfcodes** [reply]
@MichaelBurgess did you ever figure this out? I'm having the same issue.

**Reply by christiankaindl** [reply]
Yes, I also fell into this rabbit hole. When you host on Vercel the function response size is [4.5MB](searchIndex.exists()) (technically 5MB, but 0.5MB is reserved for headers). The problem is that the response from `getStaticProps` is just the page *data*, not the size of the actual HTML that is returned to the client.

For me, I did 2 things:
- Tried to reduce the HTML size by reducing unnecessary DOM elements and moving inline styles to CSS classes as much as possible
- Implemented a fallback strategy in `getStaticProps()` as follows: when the payload size is larger than a certain threshold (e.g. 2MB), I upload the page data to S3 and only return the object URL. On the client I then fetch the data and show a spinner while loading. That way small pages still get pre-rendered nicely, and large pages use the fallback strategy using S3.

Sadly there is no way of knowing how large the HTML response size will be in advance, so the payload size threshold is something that you'll need to figure out based on your page weight.

Pseudo code:

```ts
const MAX_RESPONSE_SIZE = 2500000 /* 2.5MB */
export const getStaticProps: GetStaticProps = async function ({ params }) {
  const data = await getPageData(params.id)

  // When over the size threshold, use fallback strategy
  if (Buffer.from(JSON.stringify(data)).length > MAX_RESPONSE_SIZE) {
    // Upload to storage service
    const myUrl = await uploadToService(data)
    return {
      props: { fallbackDataUrl: myUrl }
    }
  }

  // Return full data
  return {
    props: { data }
  }
}
```

Hope this helps!
